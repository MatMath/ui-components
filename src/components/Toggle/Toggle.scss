@use '../../colors/colors';
@use '../../typography/typography';

/* ROOT */
.root {
  display: flex;
  flex-direction: row;
}

/* INPUT */
input {
  border: 0;
  clip: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;

  &:focus + .groove {
    /* To be defined */
    background: darkred;
    /* box-shadow: 0 0 0 3px pink; */
  }

  &:focus + .handle {
    /* To be defined */
    background: red;
    /* box-shadow: 0 0 0 3px pink; */
  }
}

/* GROOVE */
.groove {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 32px;
  height: 16px;
  cursor: pointer;
  border-radius: 8px;
  position: relative;
  transition: background-color 0.2s;
  background: colors.$white;

  border: 1px solid colors.$dash-green-05;
  box-sizing: border-box;

  &.checked {
    border-width: 0px;
    background: colors.$mid-green-00;

    &.disabled {
      background: colors.$mid-green-03;
    }

    &.hover {
      background: colors.$dash-green-00;
    }
  }

  &.disabled {
    cursor: 'default';
    background: colors.$dash-green-06;
  }
}

/* HANDLE */
.handle {
  content: '';
  position: absolute;
  top: 'calc(100% - 4px)';
  width: 8px;
  height: 8px;
  border-radius: 50%;
  transition: 0.2s;
  background: colors.$mid-green-00;
  transform: none;
  left: 4px;

  &.checked {
    transform: translateX(-100%);
    left: calc(100% - 4px);
    background: colors.$white;
  }

  &.unchecked {
    &.disabled {
      background: colors.$mid-green-03;
    }

    &.hover {
      background: colors.$dash-green-00;
    }
  }
}

/* LABEL */
.label {
  @include typography.body-small;
  margin-left: 8px;
  color: colors.$black;
  cursor: pointer;

  &.disabled {
    color: colors.$dash-green-03;
    cursor: default;
  }
}
