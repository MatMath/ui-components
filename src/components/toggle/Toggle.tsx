import * as React from 'react';
import { getClassNames } from '@utility/cssUtils';
import styles from './Toggle.module.scss';

interface ToggleProps extends React.InputHTMLAttributes<HTMLInputElement> {
  labelUnchecked?: string;
  labelChecked?: string;
}

interface GrooveProps {
  checked?: boolean;
  disabled?: boolean;
}

interface HandleProps {
  checked?: boolean;
  disabled?: boolean;
}

interface LabelProps {
  checked?: boolean;
  labelUnchecked?: string;
  labelChecked?: string;
}

const HiddenInput: React.FC<
  React.InputHTMLAttributes<HTMLInputElement>
> = props => (
  <input className={styles.hiddenInput} type='checkbox' {...props} />
);

const Groove: React.FC<GrooveProps> = ({ checked, disabled, children }) => (
  <div
    className={getClassNames(styles.groove, {
      [styles.checked]: checked,
      [styles.disabled]: disabled
    })}
  >
    {children}
  </div>
);

const Handle: React.FC<HandleProps> = ({ checked, disabled }) => (
  <div
    className={getClassNames(styles.handle, {
      [styles.checked]: checked,
      [styles.disabled]: disabled
    })}
  />
);

const Label: React.FC<LabelProps> = ({
  checked,
  labelChecked,
  labelUnchecked
}) => {
  const isLabelCheckedVisible = checked && labelChecked;
  const isLabelUncheckedVisible = !checked && labelUnchecked;

  return isLabelCheckedVisible || isLabelUncheckedVisible ? (
    <span className={styles.text}>
      {checked ? labelChecked : labelUnchecked}
    </span>
  ) : null;
};

export const Toggle: React.FC<ToggleProps> = props => {
  const { labelChecked, labelUnchecked, ...hiddenInputProps } = props;

  return (
    <label
      className={getClassNames(styles.root, {
        [styles.checked]: props.checked,
        [styles.disabled]: props.disabled
      })}
    >
      <HiddenInput {...hiddenInputProps} />
      <Groove checked={props.checked} disabled={props.disabled}>
        <Handle checked={props.checked} disabled={props.disabled} />
      </Groove>
      <Label
        checked={props.checked}
        labelUnchecked={labelUnchecked}
        labelChecked={labelChecked}
      />
    </label>
  );
};
