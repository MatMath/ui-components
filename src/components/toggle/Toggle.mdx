---
name: Toggle
menu: Components
---

import { Playground, Props } from 'docz';
import { Toggle } from './Toggle';
import { ToggleDemo } from './ToggleDemo';

# Toggle

An On/Off checkable toggle component.

## Properties

<Props of={Toggle} />

## Usage

### Basic usage

Examples below show the `Toggle` in its checked and unchecked states, both in enabled and disabled modes:

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around',
    pointerEvents: 'none'
  }}
>
  <Toggle readOnly />
  <Toggle checked readOnly />
  <Toggle disabled readOnly />
  <Toggle disabled checked readOnly />
</Playground>

The example below shows an interactive `Toggle`.

<Playground
  style={{
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around'
  }}
>
  <ToggleDemo />
</Playground>
