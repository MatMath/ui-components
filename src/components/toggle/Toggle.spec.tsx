import * as React from 'react';
import { mount } from 'enzyme';
import 'jest-styled-components';
import { Toggle } from './Toggle';

describe('<Toggle>', () => {
  describe('Global render', () => {
    it('should render default', () => {
      const toggle = mount(<Toggle />);
      expect(toggle).toMatchSnapshot();
    });

    it('should render disabled', () => {
      const toggle = mount(<Toggle disabled />);
      expect(toggle).toMatchSnapshot();
    });

    it('should render checked', () => {
      const toggle = mount(<Toggle checked defaultChecked />);
      expect(toggle.props().checked).toEqual(true);
      expect(toggle).toMatchSnapshot();
    });

    it('should render with labelOn', () => {
      const labelOn = 'On';
      const toggle = mount(<Toggle labelOn={labelOn} checked />);
      expect(toggle.find('span').text()).toEqual(labelOn);
      expect(toggle).toMatchSnapshot();
    });

    it('should render with labelOff', () => {
      const labelOff = 'Off';
      const toggle = mount(<Toggle labelOff={labelOff} />);
      expect(toggle.find('span').text()).toEqual(labelOff);
      expect(toggle).toMatchSnapshot();
    });
  });
});
