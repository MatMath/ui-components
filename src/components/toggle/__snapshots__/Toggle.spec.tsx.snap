// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle> Global render should render checked 1`] = `
<Component
  checked={true}
  defaultChecked={true}
>
  <label
    className="root checked"
  >
    <HiddenInput
      checked={true}
      defaultChecked={true}
    >
      <input
        checked={true}
        className="hiddenInput"
        defaultChecked={true}
        type="checkbox"
      />
    </HiddenInput>
    <Groove
      checked={true}
      defaultChecked={true}
    >
      <div
        className="groove checked"
      >
        <Handle
          checked={true}
          defaultChecked={true}
        >
          <div
            className="handle checked"
          />
        </Handle>
      </div>
    </Groove>
    <Label
      checked={true}
      defaultChecked={true}
    />
  </label>
</Component>
`;

exports[`<Toggle> Global render should render default 1`] = `
<Component>
  <label
    className="root"
  >
    <HiddenInput>
      <input
        className="hiddenInput"
        type="checkbox"
      />
    </HiddenInput>
    <Groove>
      <div
        className="groove"
      >
        <Handle>
          <div
            className="handle"
          />
        </Handle>
      </div>
    </Groove>
    <Label />
  </label>
</Component>
`;

exports[`<Toggle> Global render should render disabled 1`] = `
<Component
  disabled={true}
>
  <label
    className="root disabled"
  >
    <HiddenInput
      disabled={true}
    >
      <input
        className="hiddenInput"
        disabled={true}
        type="checkbox"
      />
    </HiddenInput>
    <Groove
      disabled={true}
    >
      <div
        className="groove disabled"
      >
        <Handle
          disabled={true}
        >
          <div
            className="handle disabled"
          />
        </Handle>
      </div>
    </Groove>
    <Label
      disabled={true}
    />
  </label>
</Component>
`;

exports[`<Toggle> Global render should render with labelOff 1`] = `
<Component
  labelOff="Off"
>
  <label
    className="root"
  >
    <HiddenInput
      labelOff="Off"
    >
      <input
        className="hiddenInput"
        type="checkbox"
      />
    </HiddenInput>
    <Groove
      labelOff="Off"
    >
      <div
        className="groove"
      >
        <Handle
          labelOff="Off"
        >
          <div
            className="handle"
          />
        </Handle>
      </div>
    </Groove>
    <Label
      labelOff="Off"
    >
      <span
        className="text"
      >
        Off
      </span>
    </Label>
  </label>
</Component>
`;

exports[`<Toggle> Global render should render with labelOn 1`] = `
<Component
  checked={true}
  labelOn="On"
>
  <label
    className="root checked"
  >
    <HiddenInput
      checked={true}
      labelOn="On"
    >
      <input
        checked={true}
        className="hiddenInput"
        type="checkbox"
      />
    </HiddenInput>
    <Groove
      checked={true}
      labelOn="On"
    >
      <div
        className="groove checked"
      >
        <Handle
          checked={true}
          labelOn="On"
        >
          <div
            className="handle checked"
          />
        </Handle>
      </div>
    </Groove>
    <Label
      checked={true}
      labelOn="On"
    >
      <span
        className="text"
      >
        On
      </span>
    </Label>
  </label>
</Component>
`;
